src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp

```diff
@@ -38,13 +39,102 @@ namespace tosa {
+llvm::SmallVector<int64_t> getCeilConstants(llvm::ArrayRef<int64_t> inputShape,
+    const ONNXGenericPoolOpShapeHelper<ShapeHelperType> &shapeHelper,
+    int64_t ceilMode) {
+  // This avoids having more if statements when creating the padding const.
+  if (ceilMode == 0)
+    return llvm::SmallVector<int64_t>{0, 0};
+
+  llvm::SmallVector<int64_t, 2> kernelShapeVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.kernelShape);
+
+  // Get stride and pad vectors.
+  llvm::SmallVector<int64_t, 2> stridesVec = shapeHelper.strides;
+  llvm::SmallVector<int64_t, 4> padsVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.pads);
+
+  // Check if the idiv_check for the output dimentions in
+  // https://www.mlplatform.org/tosa/tosa_spec.html#_max_pool2d has no
+  // remainder. If it has a remainder, we add size(stride) to the end of the
+  // padding dimension to get one dimension up. Height and width need to have
+  // seperate values.
+  int64_t xAxis = 0;
+  if ((inputShape[2] + padsVec[0] + padsVec[2] - kernelShapeVec[0]) %
+      stridesVec[0])
+    xAxis = stridesVec[0];
+
+  int64_t yAxis = 0;
+  if ((inputShape[3] + padsVec[1] + padsVec[3] - kernelShapeVec[1]) %
+      stridesVec[1])
+    yAxis = stridesVec[1];
+
+  return llvm::SmallVector<int64_t>{xAxis, yAxis};
+}
+
+// Lower MaxPool and AveragePool to TOSA ops.
+template <typename ONNXPoolOp, typename TOSAPoolOp>
+llvm::Optional<mlir::Value> convertPoolOp(
+    mlir::PatternRewriter &rewriter, mlir::Operation *op) {
+  using OpAdaptor = typename ONNXPoolOp::Adaptor;
+  auto poolOp = llvm::cast<ONNXPoolOp>(op);
+  mlir::Location loc = op->getLoc();
+  OpAdaptor adaptor(op->getOperands(), op->getAttrDictionary());
+  // Get shape.
+  IndexExprBuilderForTosa createTosaIE(rewriter, loc);
+  ONNXGenericPoolOpShapeHelper<ONNXPoolOp> shapeHelper(op, {}, &createTosaIE);
+  shapeHelper.computeShapeAndAssertOnFailure();
+
+  TosaBuilder tosaBuilder(rewriter, loc);
+
+  mlir::Value input = adaptor.X();
+
+  mlir::ArrayAttr kernelShape = adaptor.kernel_shapeAttr();
+  const int64_t ceilMode = adaptor.ceil_mode();
+
+  auto inputType = input.getType().cast<mlir::TensorType>();
+
+  // Construct the transposed type for the new Pool OP
+  mlir::Type newResultType = mlir::RankedTensorType::get(
+      llvm::SmallVector<int64_t, 4>(inputType.getShape().size(), -1),
+      inputType.getElementType());
+
+  // ONNX Mlir uses NCHW as an input while TOSA expects NHWC. Insert a
+  // transpose to change the format
+  input = tosaBuilder.transpose(input, {0, 2, 3, 1});
+
+  // When ceil mode is 1, we need to add values to the padding
+  const llvm::SmallVector<int64_t, 4> ceilConstants =
+      getCeilConstants<ONNXPoolOp>(inputType.getShape(), shapeHelper, ceilMode);
+  llvm::SmallVector<int64_t, 4> pads =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.pads);
+
+  // reorder padding values
+  mlir::ArrayAttr newPads = rewriter.getI64ArrayAttr({pads[0],
```

Comment:
Nit : This will create a I64 const, and the previous call to tosaBuild.transpose actually creates a i32 const if my memory is correct. I noticed in the IR recently that we had a mix of both for constants. It's not issue, but that would be great if we we always using the same type. What do you think?

Reply:
You are right, I changed most if the implementation to i32, because that is what the TOSA Specification says. TOSA MLIR allows a bit more. I will change this to i32 to reflect the specification!

Reply:
Interesting enough, TOSA MLIR wants this to be i64 Attributes. Seems to be a departure from the standard. I will not change this then, but it is indeed strange